Assistant: Initial pass at implementing a data summary tool for Python

melissa-barca opened the pull request:
First pass at #7114 

Provides Assistant with a `getDataSummary` tool, currently only implemented for Python, that provides a JSON structured summary of a data object by using the Positron API to communicate with the Variables Comm. I updated the variable's python backend to reuse existing functionality from the data explorer. 

I used the `inspectVariables` tool as a guide for retrieving info from the variables comm.

<img width="832" alt="image" src="https://github.com/user-attachments/assets/eb2051f8-4716-4030-a685-9853cc2eb2e6" />

### Release Notes

<!--
  Optionally, replace `N/A` with text to be included in the next release notes.
  The `N/A` bullets are ignored. If you refer to one or more Positron issues,
  these issues are used to collect information about the feature or bugfix, such
  as the relevant language pack as determined by Github labels of type `lang: `.
  The note will automatically be tagged with the language.

  These notes are typically filled by the Positron team. If you are an external
  contributor, you may ignore this section.
-->

#### New Features

- N/A

#### Bug Fixes

- N/A


### QA Notes

<!--
  Positron team members: please add relevant e2e test tags, so the tests can be
  run when you open this pull request.

  - Instructions: https://github.com/posit-dev/positron/blob/main/test/e2e/README.md#pull-requests-and-test-tags
  - Available tags: https://github.com/posit-dev/positron/blob/main/test/e2e/infra/test-runner/test-tags.ts
-->
@:data-explorer
@:assistant
@:variables
@:plots
@:viewer

<!--
  Add additional information for QA on how to validate the change,
  paying special attention to the level of risk, adjacent areas that
  could be affected by the change, and any important contextual
  information not present in the linked issues.
-->

wesm commented on extensions/positron-assistant/src/md/prompts/chat/agent.md:
```diff
@@ -44,4 +59,4 @@ You adhere to the following workflow when dealing with package management:
 5. Never instruct users to install, load, or import packages that are already loaded in their session
 6. Do not generate conditional code (if/then statements) to check package availability. Use the provided tools to determine package status and generate only the necessary installation or loading code based on the tool results
-</package-management>
+   </package-management>
```
Unintended change?

wesm commented on extensions/positron-python/python_files/posit/positron/variables_comm.py:
```diff
@@ -183,6 +183,9 @@ class VariablesBackendRequest(str, enum.Enum):
 
+    # Summarize data
+    SummarizeData = "summarize_data"
+
```
We'll have to modify the OpenRPC specs so that the changes in this file can be auto-generated

wesm commented on extensions/positron-python/python_files/posit/positron/variables.py:
```diff
@@ -707,6 +712,96 @@ def _send_details(self, _path: list[str], value: Any = None):
+                    "type_display": column.type_display,
+                    "summary_stats": summary_stats,
+                }
+            )
```
It's a good starting point to have this tool surfaced in the variables comm — since computing summary stats or other computed profiles can be expensive (and thus block other messaging handling in the variables comm), we'll probably want to separate "expensive" requests (e.g. summary stats, frequency tables, histograms, etc.) from "cheap" requests (like asking for the schema), and make sure that the expensive requests and performed in an asynchronous-response pattern like the `get_column_profiles` request in the data explorer. This doesn't all have to get done in this PR so can be follow up work

wesm commented on extensions/positron-python/python_files/posit/positron/variables.py:
```diff
@@ -707,6 +712,96 @@ def _send_details(self, _path: list[str], value: Any = None):
+
+            schema = TableSchema(columns=column_schemas)
+        except Exception as e:
+            raise ValueError(f"Failed to get schema: {e}") from e
```
We should be able to push some of these tool-calling helpers into functions in data_explorer.py that we can unit test there

melissa-barca replied:
I haven't added any unit tests, but I did add `_get_column_profiles` and `_get_table_schema_from_view` as helpers to data_explorer.py

wesm commented on src/vs/workbench/api/common/positron/extHost.positron.protocol.ts:
```diff
@@ -53,6 +53,7 @@ export interface MainThreadLanguageRuntimeShape extends IDisposable {
 	$interruptSession(handle: number): Promise<void>;
 	$focusSession(handle: number): void;
 	$getSessionVariables(handle: number, accessKeys?: Array<Array<string>>): Promise<Array<Array<Variable>>>;
+	$getSessionVariableDataSummaries(handle: number, accessKeys: Array<Array<string>>): Promise<Array<string>>;
```
I was thinking that to avoid a proliferation of new APIs related to data analysis tool calling, that we could potentially add a "thicker" API here with multiple options, something like:

`querySessionVariable(handle, accessKey, queryType, queryParams)`

where for now `queryType` could be one of `"schema"`, `"summary_stats"`, etc. Then we can add other types of tools without having to touch the API plumbing each time. There can still be multiple tools but the tool calls can all get "routed" through this "thick" API. Let me know what you think

wesm reviewed (commented):
This looks like a great start. My main suggestion is to rename the API that routes requests to the variables comm to something more generic (and it can just query a single session variable at a time) so that we can use it to add more data querying tools without having to modify the Positron API each time

The other changes that we will want to make is to make the handling of these tool calls "asynchronous" so they they do not block the functioning of the variables comm — this means basically copying the pattern from the data explorer comm for the get_column_profiles request (and its corresponding return_column_profiles front-end API, see https://github.com/posit-dev/positron/blob/main/extensions/positron-python/python_files/posit/positron/data_explorer.py#L492-L519)

melissa-barca committed:
initial pass at implementing a data summary tool for Python

improve logging performance to satisfy linter

clean up code

provide temp comm to satisfy pyright

modify openRPC specs to autogen comms ccode and fix bug with passing
'path' parameter, also rename summarizeData function to make it more
generic

create data explorer helper functions

revert formatting change

wesm commented on extensions/positron-python/python_files/posit/positron/variables.py:
```diff
@@ -707,6 +711,75 @@ def _send_details(self, _path: list[str], value: Any = None):
+            # Create a temporary table view with a temporary comm
+            temp_state = DataExplorerState("temp_summary")
+            temp_comm = PositronComm.create(target_name="temp_summary", comm_id="temp_summary_comm")
+            table_view = _get_table_view(value, temp_comm, temp_state, self.kernel.job_queue)
```
Maybe later we can set up a persistent data explorer comm to use for Assistant tool calls (I realized just now after my earlier comment about the async column profiles — not needed for now — that these depend on there being a live comm available to send the frontend event though with the asynchronous result. We can look more closely at this later)

wesm commented on positron/comms/variables-backend-openrpc.json:
```diff
@@ -189,6 +189,57 @@
+					"description": "Result of the summarize operation",
+					"type": "object",
+					"properties": {
+						"children": {
```
This is returning a different return type right now (with the schema and column profiles, so a lot more complex). I think to avoid having to drag along the schema and profile result type (and mainly having to expose these in the Positron runtime / extHost API) we can just return the schema and profiles as a serialized JSON string to sidestep this issue for now -- it would be good to make these results well-typed everywhere but there's a bunch of plumbing needed).

wesm reviewed (commented):
This is close to a good stopping point for the initial pass — I think the main thing that need to get fixed is the return type for the query_variable_data RPC — since it isn't easy to access all of the data explorer comm types in all the layers where this function is called, we can just return serialized JSON from the function for now (effectively `schema: string, column_profiles: string[]`)

wesm commented:
I rebased this today and will work on some unit tests on the Python backend portion before it can be merged

wesm committed:
Rename tool as getTableSummary and pass more type information through the ext host API

wesm committed:
Simplify data explorer request handling, use common code path for serving get schema requests

wesm committed:
Work toward unit testing

wesm committed:
Add unit test to smoke test the tool RPC handler

wesm commented:
@sharon-wang @jmcphers I think I've got this to a good stopping point on the Python side — I can go ahead and merge but it will be broken for R until #8343 is tackled (shouldn't be too difficult, I don't think!). Let me know how you would like to proceed

sharon-wang commented on extensions/positron-assistant/src/tools.ts:
```diff
@@ -299,6 +299,56 @@ export function registerAssistantTools(
+				]);
+			}
+
+			// temporarily only enable for Python sessions
+			let session: positron.LanguageRuntimeSession | undefined;
+			const sessions = await positron.runtime.getActiveSessions();
+			if (sessions && sessions.length > 0) {
+				session = sessions.find(
+					(session) => session.metadata.sessionId === options.input.sessionIdentifier,
+				);
+			}
+			if (!session) {
+				return new vscode.LanguageModelToolResult([
+					new vscode.LanguageModelTextPart('[[]]')
+				]);
+			}
+
+			if (session.runtimeMetadata.languageId !== 'python') {
+				return new vscode.LanguageModelToolResult([
+					new vscode.LanguageModelTextPart('[[]]')
+				]);
+			}
```
Thoughts on filtering out this tool when there's no session and (temporarily) if it's not a python session? Maybe we can also temporarily add a note to the tool description that the tool is only available for Python. This way, the tool shouldn't be available or run at all.

We have some tool filtering logic here:

https://github.com/posit-dev/positron/blob/90b4e5ca3b459c65f692c8573424467a1e9ad220/extensions/positron-assistant/src/participants.ts#L201-L258

Otherwise, we could throw an Error noting that this is only available for Python or return a string instead of returning an empty text part, just so it's clear to the user and the model why we were unable to grab the table summary info?

wesm replied:
Agreed, let me see if I can figure out how to do that

wesm replied:
<img width="495" height="331" alt="image" src="https://github.com/user-attachments/assets/90038dfd-326e-41f6-9a20-87d8c3250c6f" />

sharon-wang commented on extensions/positron-assistant/src/tools.ts:
```diff
@@ -299,6 +299,56 @@ export function registerAssistantTools(
+
+			// If no session identifier is provided, return an empty array.
+			if (!options.input.sessionIdentifier || options.input.sessionIdentifier === 'undefined') {
+				return new vscode.LanguageModelToolResult([
+					new vscode.LanguageModelTextPart('[[]]')
+				]);
```
Should we throw an error here?

sharon-wang reviewed (commented):
I'm uninitiated with the python data explorer code, so I mostly looked at the assistant changes!

For https://github.com/posit-dev/positron/issues/8343, is the leftover work to implement the equivalent to the `extensions/positron-python` changes in this PR for `extensions/positron-r`?

wesm committed:
Fix up agent.md, exclude getTableSummary tool for non-Python sessions

sharon-wang reviewed (dismissed):
LGTM! I'm seeing the tool only available for Python sessions. Thank you for adding this 🙌

sharon-wang commented on extensions/positron-assistant/package.json:
```diff
@@ -358,6 +358,39 @@
+            "accessKeys"
+          ]
+        },
+        "tags": [
+          "positron-assistant"
+        ]
```
Sorry, one more thing before merging! Since https://github.com/posit-dev/positron/pull/8495 has gone in, we'll also need to add this tag:
```suggestion
        "tags": [
          "positron-assistant",
          "requires-session"
        ]
```

sharon-wang commented on extensions/positron-assistant/src/participants.ts:
```diff
@@ -283,6 +283,9 @@ abstract class PositronAssistantParticipant implements IPositronAssistantPartici
 						return inChatPane && (isEditMode || isAgentMode);
+					// Only include the getTableSummary tool for Python sessions until supported in R
+					case PositronAssistantToolName.GetTableSummary:
+						return positronContext.activeSession?.language.toLowerCase() === 'python';
```
Ah, I think `activeSession` was refactored out of the context in https://github.com/posit-dev/positron/pull/8495.

This change works for me locally:

```suggestion
						// TODO: Remove this restriction when the tool is supported in R https://github.com/posit-dev/positron/issues/8343
						// The logic above with TOOL_TAG_REQUIRES_ACTIVE_SESSION will handle checking for active sessions once this is removed.
						return activeSessions.has('python');
```

wesm committed:
Update extensions/positron-assistant/src/participants.ts

Co-authored-by: sharon <sharon-wang@users.noreply.github.com>
Signed-off-by: Wes McKinney <wesm@apache.org>

sharon-wang reviewed (approved):
Yay checks are green!
